Add details on native packaging requirements exposed by mobile platforms (#27)
pypackaging-native/pypackaging-native: docs/key-issues/multiple_architectures.md

# Platforms with multiple CPU architectures

In addition to any ABI requirements, a binary is compiled for a CPU
architecture. That CPU architecture defines the CPU instructions that can be
issued by the binary.

Historically, it could be assumed that an executable or library would be
compiled for a single CPU archicture. On the rare occasion that an operating
system was available for mulitple CPU architectures, it became the
responsibility of the user to find (or compile) a binary that was compiled for

> their host CPU architecture.

Reviewer: I think this article focuses too much on what (from my experience; away from the mobile ecosystem) is a niche use-case. Linux is very commonly compiled on all kinds of architectures, and the simple fact is that the architecture _needs_ to be taken into account (see "target triple" also in the glossary).

The `universal2` wheels are an exception due to Apple bridging the gap of their architecture change.

What's a common collection of architectures a binary needs to support on Android?

Author: Answer in reverse order: On Android, it's common to need armeabi-v7a, arm64-v8a, x86 and x86-64. The former two are 32 and 64 bit ARM, respectively; that's primarily for usage on device, but on macOS you'll need those for the simulator as well. The x86 and x86-64 images are entirely for simulator usage.

As for the other point: architecture is indeed important - each "slice" in a fat binary needs to be for the correct architecture. Compiler triples in particular are one of the biggest headaches in supporting multi-architecture builds. After all - what is the compiler triple for macOS supporting x86 *and* arm64? FWIW, Python's build system essentially ignores the architecture part of the compiler triple once it works out it's on macOS; when building third party libraries, you are essentially forced into the "lipo" solution I mention below - build multiple single-platform binaries, and then merge them into a single binary.

However, writing this use case off as "niche" is essentially saying "we don't consider mobile or macOS platforms to be in scope". 

On mobile platforms, multiple architecture *is* the use case. True, mobile usage isn't a big part of Python right now - but I'd argue that's a "chicken and egg" problem - nobody ships binaries for mobile platforms, so nobody uses Python on mobile, so nobody ships binaries for mobile platforms. And it's not easy for end-users to resolve this themselves - due largely to the fact that Python's binary build tooling doesn't deal well with multiple architectures and cross platform builds. However, in terms of the broader market, phones and tablets have market penetration that dwarfs that of desktop platforms. I would argue we ignore mobile platforms out our peril.

As for macOS: macOS spent 2005-2012 in a the universal1 era; The universal2 era started in 2019, so it seems reasonable to assume it will persist until ~2026. However, in between those two times, there was the x86/x86-64 paring, and it is possible to have a binary that supports both of *those* architectures simultaneously. It is also possible (but uncommon) to have a "really fat" binary that supports PPC, x86, x86-64, and M1 in a single binary. M2 is architecture compatible with M1, but it's not inconceivable that M3+ may require another period of fat binaries.

I don't think the comparison with Linux is entirely reasonable. Linux is essentially the "historical" case in this paragraph - the assumption is that the end user will download code and compile it for *their* computer; if they distribute code to another user, that user will compile it for *their* computer. *My* computer only ever has a single architecture, so I only ever need to compile for *my* architecture; and you can generally assume that the end user has (or can easily obtain) a compiler for their architecture. The idea of distributing binaries is essentially alien to the general Linux experience; you get binaries from your operating system's repository, and manylinux has established a set of specific conditions to allow sharing of Python module binaries, but outside those limited subsets, you hit limitations *very* quickly if you try to share a binary between two users. 

It's distribution that is the key factor here. The Linux use case essentially doesn't involve distribution. This isn't the case for Windows and macOS - Windows and macOS users *expect* to be able to download binaries, have them "just work", and share those binaries with their co-worker. The same is true of mobile platforms. It's possible to live your entire life without ever downloading a compiler for Windows or macOS; and you *can't* install a compiler on your mobile device.

Reviewer: Thanks for the detailed reply!

> However, writing this use case off as "niche" is essentially saying "we don't consider mobile or macOS platforms to be in scope".

This is entirely due to my ignorance of that space (even though I _did_ see your 2019 PyCon keynote), and I'm not trying to write it off, as niche or otherwise (which is what I tried to caveat with "but maybe I don't know the mobile story well enough. Perhaps you can enlighten me further?").

That being said, fat binaries still feel like a hack to me. If the metadata and binary distribution were sufficiently good, you could just distribute the binary that a given installation requires, rather than duplicating (or worse) the footprint & bandwidth because the metadata isn't good enough.

I'll grant that one might view linux as the exception here, but it's kinda exceptions all the way down. Windows didn't really change architecture for a while (they tried their level best to hide x86 vs x86-64 from the user), and I don't think they're endorsing fat builds now that win-arm64 rolls around. OSX made a big bet to change architectures (that seems to pay off), but I don't see them eager to repeat that soon (perhaps I'm wrong). Mobile is it's own potpourri of history that you know way better than I do.

In any case, from the POV of describing the problems of the current ecosystem (the stated goal of pypackaging-native), it seems to me that fat binaries are a symptom (that can/should be explained) of an underlying problem (e.g. bad metadata, lack of architecture introspection of installers, etc.), but I don't (yet?) see why they necessarily need to be part of the solution space at all.[^1]

[^1]: it's of course possible that my idealistic distribute-only-what-you-need runs into real world issues and fat binaries might still come out as the least bad trade-off. But we're a ways off from the ecosystem-level discussion getting into that sort of nitty gritty.

Reviewer 2: > The idea of distributing binaries is essentially alien to the general Linux experience; you get binaries from your operating system's repository, and manylinux has established a set of specific conditions to allow sharing of Python module binaries, but outside those limited subsets,

The idea of binaries from an operating system's repository being a "limited subset" strikes me as somehow highly strange.

And it definitely doesn't take into account the many, many, many projects out there that provide their own debian and Fedora repositories, which are architecture-bound and contain single-architecture packages that don't really care what actual version of either distro you are running. Or else are just compiled for the latest stable release -- that works too.

This is effectively identical to the manylinux specification, except manylinux is a sort of distro repository that works on many distros, but not nearly as well as a .deb or .rpm work.

Linux distro repositories are the original appstore, it just so happens that there's less fragmentation for Android and iOS -- while both have many versions, there's only the two platforms, not several dozen, so it's easy to package for either one.

It's also not a concept that Windows has missed out on noticing. Chocolatey is justifiably popular, and Microsoft based winget on that, sort of. Both allow easily downloading and installing existing architecture-specific *.msi installers etc. based on whichever architecture you are running.

...

There's only one key distinction to make between appstores for:
- Apple 
- everything else

Linux and Android both require vendors (whether the vendor is a distro or a project does not matter) to compile once for each architecture, and upload multiple binaries. Those multiple binaries, when ingested into an appstore, are abstracted away into the marketing name (package name or whatever), users ask to install that, and they get whichever one the appstore determines is best for them.

Apple requires:
- vendors to compile a single binary that is twice as big
- users to download a single binary that is twice as big
- users to store a single binary that is twice as big

And frankly, none of it is necessary at all.  A `*.pkg` should either specify the architecture it is for, or contain multiple architectures with each architecture tagged as relevant. For "universal2" packages, the appstore could perform the slicing for you, and only deliver the bits you need repackaged into a new .pkg, or simply work the way the appstore work like appstores on other platforms -- just deliver the one you need.

There is theoretical value in hosting the .pkg as a single file that users download and click on, although the only reason this value exists is because Apple doesn't like people to have control of their systems and offering custom appstore indexes makes Apple lose control. Even in this case, however, the installation process of that package should strip out what you do not need and avoid burning disk space frivolously. Granted that network bandwidth has already been frivolously wasted. :(

> However, writing this use case off as "niche" is essentially saying "we don't consider mobile or macOS platforms to be in scope".

What about option 3:

mobile or macOS platforms are in scope, but mobile platforms other than Apple aren't the niche under discussion, and Apple can be simultaneously in scope but also be creating its own unnecessary problems which is grumble-worthy. :)

Reviewer 2: I guess the takeaway from this is, this article feels to me like:
- it de-emphasizes the state of the art in the *normative* case of "an operating system platform where multiple CPU architectures are supported" (which is nearly all platforms that are recent or even not-so-recent)
- the Android approach is not novel to Android -- it's the state of the art on Linux and the BSDs, and predates Android, but that's being ignored
- the PR is for "requirements exposed by mobile platforms" but it talks about "multiple architectures" which isn't the same at all, and then spends quite a bit of time talking about non-mobile platforms as "rare" or "on occasion" (?)

Author: > That being said, fat binaries still feel like a hack to me. If the metadata and binary distribution were sufficiently good, you could just distribute the binary that a given installation requires, rather than duplicating (or worse) the footprint & bandwidth because the metadata isn't good enough.

I won't argue they're a hack - but FWIW, they're a moderately elegant one. It is 100% a tradeoff of disk space for user experience. As a macOS user, I can "just download a binary", and it will "Just Work". I don't need to know whether I have an x86-64 or M1 machine, and I can't download the "wrong" binary for my machine.
 
From a developer's POV, it's also pretty elegant... as long as you're in the macOS ecosystem. In simple cases,  replacing `clang` with `xcrun --sdk macosx clang` will generate a fat binary, and you're done. Compilations only start to emerge when you hit against code that doesn't have a macOS story baked in. For example, if a project has a makefile that decomposes the compiler triple and tries to determine something significant from the architecture; or expects a `-D` flag to set the architecture, you're going to have a Bad Time. For these projects, you usually need to fall back on building multiple versions of the library and then mashing them together with `lipo`. It's not impossible to do, it's just fiddly, and most projects don't have a good place for "multiple pass then mash together" build logic to live, so you end up with meta-makefiles or similar.

> I'll grant that one might view linux as the exception here, but it's kinda exceptions all the way down. Windows didn't really change architecture for a while (they tried their level best to hide x86 vs x86-64 from the user), and I don't think they're endorsing fat builds now that win-arm64 rolls around. OSX made a big bet to change architectures (that seems to pay off), but I don't see them eager to repeat that soon (perhaps I'm wrong). Mobile is it's own potpourri of history that you know way better than I do.

I wouldn't take bets on another big platform shift any time soon; but I *would* take bets on a new iteration of ARM. ARM64 is really "ARMv8", and "ARMv9" has been announced. Apple hasn't announced anything about ARMv9 (AFAIK), but the fat approach allows them to support both in a single binary. On iOS, in the recent past it was necessary to support ARMv6, ARMv7, ARMv7s and ARM64 in a single binary if you wanted to support all Apple-supported devices.

> In any case, from the POV of describing the problems of the current ecosystem (the stated goal of pypackaging-native), it seems to me that fat binaries are a symptom (that can/should be explained) of an underlying problem (e.g. bad metadata, lack of architecture introspection of installers, etc.), but I don't (yet?) see why they necessarily need to be part of the solution space at all.[1](#user-content-fn-1-be6d2ab6e89d02da1a0841cf288f03f8)
> 
> ## Footnotes
> 1. it's of course possible that my idealistic distribute-only-what-you-need runs into real world issues and fat binaries might still come out as the least bad trade-off. But I think we're a ways off from the ecosystem-level discussion getting into that sort of nitty gritty. [↩](#user-content-fnref-1-be6d2ab6e89d02da1a0841cf288f03f8)

I don't agree that fat binaries are a symptom of a problem; I'd argue they're a solution to a problem that prioritises user experience over disk space (and possibly original developer experience).

As for why I think this should be in scope: IMHO, its a distribution problem - but a different distribution to the one that pip, conda, et al have historically been concerned with. 

Pip et al have historically considered "distribution" as a problem of a developer distributing a Python library to a Python package repository so that an end user can download that library into their virtualenv/conda env etc. In that context, I agree that multi-architecture problems don't exist. If I have an M1 macOS machine, I download the arm64 binary, and all is good with the world. 

The complication then comes when I, as a user who downloaded the M1 machine, try to share *my* project with my colleague, who has an x86-64 macOS machine. My single-platform binaries won't work on their machine; but a fat universal2 binary will. 

You could argue that I'm "doing it wrong" - that I should be documenting a requirements file, and my colleague should download their own set of requirements appropriate for their platform. However, that requires that my colleague understand the full ecosystem for the code. I can't just throw something onto a thumb drive and ask them to run it. For want of a term to hang this on, lets call this "app distribution".

In the mobile case, this is the *only* distribution mode that is possible. If you write an app that tries to download from PyPI/conda-forge, it will be rejected by the App Store (the Apple one, anyway; Play store might be a bit more lenient).

Author: > > The idea of distributing binaries is essentially alien to the general Linux experience; you get binaries from your operating system's repository, and manylinux has established a set of specific conditions to allow sharing of Python module binaries, but outside those limited subsets,
> 
> The idea of binaries from an operating system's repository being a "limited subset" strikes me as somehow highly strange.

This is literally what the manylinux definition is. A manylinux package may only link against a handful of binaries that are guaranteed to exist.

> And it definitely doesn't take into account the many, many, many projects out there that provide their own debian and Fedora repositories, which are architecture-bound and contain single-architecture packages that don't really care what actual version of either distro you are running. Or else are just compiled for the latest stable release -- that works too.

Sure. However, at that point, you're defining compatibility with "Fedora" or "Ubuntu" rather than a more nebulous "Linux". That approach works for users of those platforms, and I've definitely seen projects do that. But users of Mint, Arch, Manjaro, or any other distributions are out of luck (or, at least, will have highly variable luck); and I can tell you from my experience maintaining BeeWare that this long tail exists, and is deeply disappointed when your binary doesn't support their chosen niche Linux distro.
 
> There's only one key distinction to make between appstores for:
> 
> * Apple
> * everything else
> 
> Linux and Android both require vendors (whether the vendor is a distro or a project does not matter) to compile once for each architecture, and upload multiple binaries. 

Incorrect. As of August 2021, Google Play store requires the upload of a single, multi-platform AAB. That AAB can be constructed from individual platform-specific APKs, but the artefact you upload *must* be multi-platform.

> And frankly, none of it is necessary at all.

Says you. Apple has made a tradeoff between disk space and user experience. As a macOS user, I have no complaints about this tradeoff.

> There is theoretical value in hosting the .pkg as a single file that users download and click on, although the only reason this value exists is because Apple doesn't like people to have control of their systems and offering custom appstore indexes makes Apple lose control. 

I won't dignify this sort of trolling with a response.

Reviewer: > You could argue that I'm "doing it wrong" - that I should be documenting a requirements file, and my colleague should download their own set of requirements appropriate for their platform. However, that requires that my colleague understand the full ecosystem for the code. I can't just throw something onto a thumb drive and ask them to run it. For want of a term to hang this on, lets call this "app distribution".

I've made the [argument](https://discuss.python.org/t/python-packaging-strategy-discussion-part-1/22420/49) on discourse that python needs an integration layer -- which would be the place to solve that, such that your colleague has nothing to do, and you (ideally) not much more.

I get the appeal of the story "I just want to give my colleague something to run", but I'm not convinced that we cannot come up with a better story than just mashing arches together because we don't know what platform we might end up running on.

> In the mobile case, this is the only distribution mode that is possible. If you write an app that tries to download from PyPI/conda-forge, it will be rejected by the App Store (the Apple one, anyway; Play store might be a bit more lenient).

Is this a question of artifact signatures (also required on M1) or because nothing can get installed on iOS that hasn't been signed off by Apple? For the former we'd have an answer (well, a hack) at least.

On a metalevel, I am not very sympathetic to improving Apple's ecosystem on a volunteer basis in the face of very deliberate hurdles they place in front of developers (and I say that as someone who has done just that for rolling out osx-arm support for many packages in conda-forge, often way before the original libraries supported that).

I understand the centrality of this stance in their business model, but I don't feel the wider community owes Apple anything in that regard. I'd much rather let them have an inferior product (for lack of a vibrant ecosystem), than mis-design our approach due to a platform that's outright hostile to what we're trying to achieve (i.e. distribute apps/binaries to users)[^1].

> I don't agree that fat binaries are a symptom of a problem; I'd argue they're a solution to a problem that prioritises user experience over disk space (and possibly original developer experience).
>
> As for why I think this should be in scope: IMHO, its a distribution problem - but a different distribution to the one that pip, conda, et al have historically been concerned with.

Changing tack a bit, how do other language ecosystems deal with distribution on phones / iOS, and can we learn from them? The scope of packaging problems is gigantic[^2] even before we add better coverage of nascent distribution channels.

Perhaps it's my biases coming through, but so far my impression is that we might want to have a mobile-specific key-issue page, without having mobile-heavy-but-not-generally-applicable pages about cross-compilation & fat binaries.

[^1]: "but what about the poor users that don't know any better?" -- over time, they always have the choice to vote with their dollars

[^2]: already just form the POV of the for (for) predominantly non-mobile set of users, which are at least as important to keep as it is to reach new ones.

Author: > > In the mobile case, this is the only distribution mode that is possible. If you write an app that tries to download from PyPI/conda-forge, it will be rejected by the App Store (the Apple one, anyway; Play store might be a bit more lenient).
> 
> Is this a question of artifact signatures (also required on M1) or because nothing can get installed on iOS that hasn't been signed off by Apple? For the former we'd have an answer (well, a hack) at least.

It's not a signing issue (although those also exist); it's a "platform security" issue. iOS App Store review policies require that all code that will be executed by the app are either (a) shipped by the app, or (b) entered by the user. You can't download code from an independent source that hasn't been through the App Store vetting process. 

This is essentially a compromise that allows app developers to use scripting languages in their own app, but doesn't allow developers to do an end-run around the App Store review process. Without this restriction, an app could essentially replace it's own running content, and Apple would not be able to verify that the app wasn't malicious. 

You can certainly argue whether this is a good or bad thing, or whether App Store review actually catches malicious behavior; but regardless, it's the current state of play.

> I understand the centrality of this stance in their business model, but I don't feel the wider community owes Apple anything in that regard.

Well, that's certainly a position one could maintain. However, I'd suggest that declaring 30% of the worldwide mobile ecosystem and 15% of the worldwide desktop ecosystem (numbers which jump to 60% and 30% respectively in the US and similar western nations) as "out of scope" because Apple doesn't conform to one specific model of software distribution will have consequences for the Python ecosystem as a whole.
 
> Changing tack a bit, how do other language ecosystems deal with distribution on phones / iOS, and can we learn from them? 

I will admit that I haven't done a *huge* amount of research, so I might be missing something; but my broad impression is that "they don't". The iOS and Android ecosystems both have very strong biases to their own languages (Swift and Kotlin, respectively). 

It's probably worth some additional research to make sure we're not missing some obvious prior art; but I wouldn't be at all surprised to discover that we're in uncharted territory.
 
> Perhaps it's my biases coming through, but so far my impression is that we might want to have a mobile-specific key-issue page, without having mobile-heavy-but-not-generally-applicable pages about cross-compilation & fat binaries.

A counterargument (framed in hyperbolic terms) - why are the concerns of GPU compatibility, which are used by a bare handful of packages in one very specific programming niche, more important than the concerns of an ecosystem that is, by some measures, the majority of worldwide computing devices?

I would not for a second claim that GPU compatibility considerations aren't important - they absolutely are. But I'd suggest that they're just part of the rich tapestry of issues facing a binary compilation and packaging ecosystem. IMHO, cross compilation and multi-architecture support are 2 other aspects of that tapestry. I would suggest treating mobile computing as a niche does the platform a disservice - especially when the issues that are being raised *aren't* mobile specific - they're also present on desktop macOS.

Reviewer: > > I understand the centrality of this stance in their business model, but I don't feel the wider community owes Apple anything in that regard.
> 
> Well, that's certainly a position one could maintain. [...]

I argue it's Apple declaring itself out of scope for many reasonable approaches. Other platform manage without their lockdowns, while Apple wants to have more control (and provides their users with a cleaner experience for it). Redhat and several others are investing a lot into the broader open source ecosystem (and are highly interoperable for it); in contrast, Apple barely does (aside from something critical to them like clang & swift).

Not that Apple's stance is morally of philosophically wrong, but it comes with consequences, and the world is not static. If Apple's approach does not support the next big thing[^1] and (say) Android does, market forces will see to it that they change tack. The key point is that (in my personal opinion), volunteer-driven organisations do not have to bend over backwards for the vagaries of such a vendor.[^2]

[^1]: In a way that whichever maintainers are willing/able to do _for_ Apple
[^2]: As an IMO similar example, nVidia is starting to pivot towards a more open(-source) approach because it became too costly for various reasons to maintain their previous stance (that made it very hard for anyone trying to work with CUDA, though many people did nevertheless because the performance gains made it worth it for them).

> This is essentially a compromise that allows app developers to use scripting languages in their own app, but doesn't allow developers to do an end-run around the App Store review process.

I don't see (for example) how their review process is engraved in stone tablets, or how it couldn't include approved channels like an official python integration layer (which is what Palantir's Foundry and Snowflake's Snowpark do when they let Python into their closed systems through a curated entry point).

Reviewer: Apple aside...

> I will admit that I haven't done a _huge_ amount of research, so I might be missing something; but my broad impression is that "they don't". The iOS and Android ecosystems both have very strong biases to their own languages (Swift and Kotlin, respectively).

This is kind of my point. Don't get me wrong - I think the work you and others are doing is absolutely necessary for the future, but solving everything at once is really hard or even impossible (and that's just for existing problems of the majority case _today_). So while the mobile case needs to be solved (and we shouldn't gratuitously throw stones in that way), I think that the addition of "blazing new trails"[^1] on top of a huge overhaul is... a bit gratuitous.

[^1]: for non-mobile-native languages

> A counterargument (framed in hyperbolic terms) - why are the concerns of GPU compatibility, which are used by a bare handful of packages in one very specific programming niche, more important than the concerns of an ecosystem that is, by some measures, the majority of worldwide computing devices?

So I agree with this, but how does that calculus look if you look at the percentages of "current Python users", or "current Python library maintainers". This is what I meant with [paraphrasing] "solving problems problems for the existing users is at least as important (probably more) than reaching new ones".

But I get that we have very divergent viewpoints on this, and there's nothing wrong with that. But you can perhaps take me as a stand-in for the average python maintainer, who (in all likelihood) has never packaged anything for mobile before, and so far, the documents you've written (haven't checked the new commits yet) read like a parallel world to me. Maybe we can agree on a common foundation and then go in more detail about challenges (mobile vs. ...) in a separate section resp. page.

Author: > The key point is that (in my personal opinion), volunteer-driven organisations do not have to bend over backwards for the vagaries of such a vendor.

Ironically, this is almost exactly my position when it comes to Linux binary support. I cannot express the size of the headaches that I've had trying to support interoperable Linux binaries with AppImage and Flatpak, or even getting manylinux binaries to work reliably.

Volunteers don't *have* to do anything. But at some point, you have to go where the users are. 

> I don't see (for example) how their review process is engraved in stone tablets, 

It's not engraved in stone - but it's not really something open for community debate or consultation. The clause that allows scripting languages at all was, for example, an addition that was made in response to industry feedback. 

I guess I'd say that an architectural choice that is based on "we'll get Apple to change their App Store review policy" is... not a viable starting position. 

> or how it couldn't include approved channels like an official python integration layer (which is what Palantir's Foundry and Snowflake's Snowpark do when they let Python into their closed systems through a curated entry point).

This sounds like an approach that would address Apple's concerns at a *technical* level; however, I'm not sure it's likely from a political/business standpoint - at least, not without someone doing a high-level strategic play with Apple at a corporate level.

> So while the mobile case needs to be solved (and we shouldn't gratuitously throw stones in that way), I think that the addition of "blazing new trails" on top of a huge overhaul is... a bit gratuitous.

Fair; although I'd argue that it *isn't* a new trail - it's a trail that has prior art for macOS support. Mobile exposes some very specific manifestations of these problems, but they're not new (or mobile specific) problems.

> So I agree with this, but how does that calculus look if you look at the percentages of "current Python users", or "current Python library maintainers". This is what I meant with [paraphrasing] "solving problems problems for the existing users is at least as important (probably more) than reaching new ones".

Can't argue with this. I'm definitely working in the "build it and they will come" space. 
 
> But I get that we have very divergent viewpoints on this, and there's nothing wrong with that. But you can perhaps take me as a stand-in for the average python maintainer, who (in all likelihood) has never packaged anything for mobile before, and so far, the documents you've written (haven't checked the new commits yet) read like a parallel world to me. Maybe we can agree on a common foundation and then go in more detail about challenges (mobile vs. ...) in a separate section resp. page.

I guess what I'm not seeing is why you consider it a parallel world, or why it's mobile specific. These problems exist *today* for macOS on desktop, and it's only by ignoring a major use case (i.e., restricting yourself to the "user downloads all their own dependencies") that these problems can be avoided.

Reviewer 2: > This is literally what the manylinux definition is. A manylinux package may only link against a handful of binaries that are guaranteed to exist.

Not sure what the manylinux definition is doing here.

You're saying that: "binaries from your operating system's repository, and manylinux" are a "limited subset" outside of which things don't work, but they're actually just the model by which it does work generically.

> Sure. However, at that point, you're defining compatibility with "Fedora" or "Ubuntu" rather than a more nebulous "Linux". That approach works for users of those platforms, and I've definitely seen projects do that. But users of Mint, Arch, Manjaro, or any other distributions are out of luck (or, at least, will have highly variable luck);

Well, no. That .deb repository works on Linux Mint too -- Mint is .deb compatible, besides which Mint is a respin of Ubuntu. Add an Arch-compatible repo, and that makes only 3 packaging formats you need to target to cover pretty much everything. The holdout is source-based distros and musl libc-based ones.

And none of that is an issue with Windows for WinGet. One format works on all Windows OSes, after selecting the correct arch-specific download in the appstore manifest.

> Incorrect. As of August 2021, Google Play store requires the upload of a single, multi-platform AAB. That AAB can be constructed from individual platform-specific APKs, but the artefact you upload must be multi-platform.

The difference between requiring that one .deb needs to be uploaded per architecture, and requiring that one APK must be uploaded per architecture "but also stick it inside of a single bundle file during the upload process, we'll split it up on our end" is not impactful.

It's still compiling once per architecture, and uploading multiple binaries.

> Says you. Apple has made a tradeoff between disk space and user experience. As a macOS user, I have no complaints about this tradeoff.

:shrug: This is a very first-world perspective. And again, it's not a trade-off between disk space and user experience. It's a trade-off between "both disk space and internet speed/availability" and "user experience for the specific subset of macOS users that 'try to share my project with my colleague'".

It produces e-waste by default, disadvantages people who live in parts of the world where the internet is slow/metered, and doesn't have any meaningful benefit to the vast majority of macOS users, who do not share development projects but only download content from an appstore. Apple claims to be very interested in being carbon-neutral, but what does this actually mean when they gratuitously waste resources?

This may be the world we live in, it may be something people in release management have to know about, but it's really not a solution or an ideal world, and if we can find something better we should.

> Well, that's certainly a position one could maintain. However, I'd suggest that declaring 30% of the worldwide mobile ecosystem and 15% of the worldwide desktop ecosystem (numbers which jump to 60% and 30% respectively in the US and similar western nations) as "out of scope" because Apple doesn't conform to one specific model of software distribution will have consequences for the Python ecosystem as a whole.

If the desktop ecosystem can just use "inferior" wheels that are x64-specific, or arm-specific, is it really declaring the platform out of scope? What are the percentages for people on desktop that actually make use of fat binaries in a way that isn't fulfilled by regular binaries which pip cleverly selects? Still the same 15% of all desktop users?

> So while the mobile case needs to be solved (and we shouldn't gratuitously throw stones in that way), I think that the addition of "blazing new trails"[1](https://github.com/pypackaging-native/pypackaging-native/pull/27#user-content-fn-1-2970fa6738cc0985f972fcfb368e10dc) on top of a huge overhaul is... a bit gratuitous.

This. 100%

However, on occasion, we see an operating system platform where multiple CPU
architectures are supported:

* In the early days of Windows NT, both x86 and DEC Alpha CPUs were supported
* Although Linux started as an x86 project, the Linux kernel is now available a
  wide range of other CPU architectures, including ARM64, RISC-V, PowerPC, s390
  and more.
* Apple transitioned Mac hardware from PowerPC to Intel (x86-64) CPUs, providing
  a forwards compatibility path for binaries
* Apple is currently transitioning Mac hardware from Intel (x86-64) to
  Apple Silicon (ARM64) CPUs, again providing a forwards compatibility
  path
* Apple supports ARMv6, ARMv7, ARMv7s, ARM64 and ARM64e on iOS
* Android currently supports ARMv7, ARM64, x86, and x86-64; it has historically

>   also supported ARMv5 and MIPS

Author: Added a note on this.

distribution, an installer or other downloadable artefact is provided for each
supported platform, and it is up to the user to select and download the correct
artefact.

### Archiving

The approach taken by Android is very similar to the multiple binary approach,
with some affordances and tooling to simplify distribution.

When building an Android project, each target architecture is compiled
independently. If a native binary library is required to compile the Android
application, a version must be provided for each supported CPU architecture. A
directory layout convention exists for providing a binary for each platform,
with the same library name. This yields an independent final binary (APK) for
each CPU architecture. When running locally, a CPU-specific APK will be
uploaded to the simulator or test device.

To simplify the process of distributing the application, at time of publication,
a single Android App Bundle (AAB) is generated from the multiple CPU-specific
APKs. This AAB contains binaries for all platforms that can be uploaded to an
app store.

When an end-user requests the installation of an app, the app store strips out the

> binary that is appropriate for the end-user's device.

Author: In the Android case, it's not really a *problem* per se; its more of a "consideration to keep in mind".

Let's say I have an Android project that uses a Python package A that has a binary module; that package has a dependency on B (which also has a binary module). When I do the compilation pass for ARM64, I `pip install A`, which does a dependency resolution pass to determine a compatible version of B. I then do a compilation pass for ARMv7, which does a separate dependency resolution pass to determine a compatible version of B. The use of a separate dependency resolution pass introduces the possibility that my ARMv7 and ARM64 binaries have entirely different versions of package B (and, I guess, potentially package A as well). 

That's not a problem; it could even be argued as "working as designed". It's how BeeWare (strictly, the Chaquopy subsystem that Briefcase uses) is working at present; it just might be a little surprising if you don't have a systematic way of at least flagging that there are different versions installed for each platform. I'll modify the wording to clarify this.

However, for the same project on iOS, this isn't an option. iOS apps perform one compilation pass per ABI (i.e. one for the simulator, and one for the physical device), so the final installed artefact needs to contain a single fat binary with all the architectures. This *could* be treated as an install-time problem rather than distribution problem, though; I'll add a note about that possible approach.

> binary that is appropriate for the end-user's device.

Reviewer 3: Thanks, this is a good and concrete example. It sounds like it's only a problem with `pip`/PyPI - any other package relevant package manager gets all the metadata first and then does a single solve for all dependencies, so you'd get a single version of package B.

> binary that is appropriate for the end-user's device.

Author: I'm not sure I follow why it's only a pip/PyPI problem. The issue isn't the solution path; it's how many times you need to run the installer (and thus the solver). If I solve and install for arm64, then solve and install for x86-64, that's 2 independent solutions, and any inconsistency in availability of packages for arm64 and x86-64 will result in different solutions. AFAICT, this will still be an issue for conda, as the issue isn't the availability of metadata for a single platform; it's the availability of the solution arrived at by a completely independent solver pass. 

You can only avoid this problem if you do a single resolver pass looking to satisfy *both* architectures at the same time, and only install once a single solution that works for both architectures is found (or, I guess, you have some ways to pass a specific solution found by pass 1 into subsequent passes in a way that *doesn't* include package hashes, as you'll be installing a different package with the same name and version, but different ABI/architecture).

> binary that is appropriate for the end-user's device.

Reviewer 3: Ah okay, thank makes sense, thanks.

> binary that is appropriate for the end-user's device.

Reviewer 2: > as the issue isn't the availability of metadata for a single platform; it's the availability of the solution arrived at by a completely independent solver pass.

conda metadata is available offline, unlike pip where every resolver run is a series of API calls followed by downloading artifacts followed by extracting artifacts followed by yet more API calls. So you actually *can* guarantee multiple runs of the resolver will produce the same results.

If the metadata is guaranteed to be consistent across architectures at any given moment (e.g. packages of any given version will always exist for all architectures) then you can also get the same resolver result across architectures too.

> or, I guess, you have some ways to pass a specific solution found by pass 1 into subsequent passes in a way that _doesn't_ include package hashes

pip freeze?

> binary that is appropriate for the end-user's device.

Reviewer 3: > So you actually _can_ guarantee multiple runs of the resolver will produce the same results.

In principle, yes. But I don't think either `conda` or `mamba` have this capability, so in practice it'd be pretty hard today.

user tries to run an binary that doesn't contain an ARM64 slice, but *does*
contain an x86-64 slice, the x86-64 slice will be converted at runtime into an
ARM64 binary. Complications can occur when only *some* of the binary is being
converted (e.g., if the binary being executed is fat, but a dynamic library
isn't).

iOS has an additional complication of requiring support for mutiple *ABIs* in
addition to multiple CPU archiectures. The ABI for the iOS simulator and
physical iOS devices are different; however, ARM64 is a supported CPU
architecture for both. As a result, it is not possible to produce a single fat
library that supports both the iOS simulator and iOS devices. Apple provides an
additional structure - the `XCFramework` - as a wrapper format for packaging
libraries that need to span multiple ABIs. When developing an application for
iOS, a developer will need to install binaries for both the simulator and
physical devices.

## Potential solutions or mitigations

Python currently provides `universal2` wheels to support x86_64 and ARM64 in a
single wheel. This is effectively a "fat wheel" format; the `.dylib` files

> contained in the wheel are fat binaries containing both x86_64 and ARM64 slices.

Reviewer 3: I would say that `universal2` should go in the "current state" rather than "potential solutions" section. Also, I'd like to point out that `universal2` is not a good idea for Python packages - its introduction seemed to be motivated mainly by "this is easiest and we're in a hurry", and the complications and disk space / bandwidth costs it entails are in general not justified (certainly not for wheels on PyPI).

there may be some overal with the solutions that can be used to support GPUs in
native binaries.

## Platform approaches for dealing with multiple architectures

Three approaches have emerged for handling multiple CPU architectures.

### Multiple binaries

The minimal solution is to distribute multiple binaries. This is the approach
that was used by Windows NT, and is currently supported by Linux. At time of
distribution, an installer or other downloadable artefact is provided for each
supported platform, and it is up to the user to select and download the correct
artefact.

### Archiving

The approach taken by Android is very similar to the multiple binary approach,
with some affordances and tooling to simplify distribution.

When building an Android project, each target architecture is compiled
independently. If a native binary library is required to compile the Android
application, a version must be provided for each supported CPU architecture. A
directory layout convention exists for providing a binary for each platform,
with the same library name. This yields an independent final binary (APK) for
each CPU architecture. When running locally, a CPU-specific APK will be

> uploaded to the simulator or test device.

Reviewer 4: The Android native compilers do indeed build each architecture independently, and produce a separate set of .so files for each one. However, it's not correct that the developer builds a separate APK for each architecture. In fact, an APK, like an AAB, contains a directory structure with a subdirectory for each architecture. And before AABs were introduced, the standard way of releasing an Android app was with a multi-architecture APK.

So with Chaquopy, the APK build process contains two steps where native code is involved:
* The standard Android support for building native libraries. Because Android is based on Java/Kotlin, most apps don't use this step.
* Chaquopy's execution of pip, as detailed below. This works with pre-compiled wheels containing .so files, and never executes any compilers.

These two build steps are independent: one does not happen within the other. And each one has its own "for each ABI" loop.

Author: Ah - thanks for the clarification. Not sure where I got mislead about APKs being single binary - I seem to recall a historical workflow with Google Play Store that involved multiple APK uploads, but I might be (a) misremembering, or (b) have been doing some redundant uploads at the time.

This approach can be supported with a conventional "single platform wheel"
approach. A library developer can package a wheel for each Android CPU
architecture they wish to support; the Android project will install a
CPU-architecture appropriate wheel when the compiler pass for that archictecture
is performed. The only complication is that process of installing wheels will
involve a dependency resolution pass on each supported platform; this could
potentially lead to a situation where a single application has different
versions of a Python library on different architectures.

To simplify the process of distributing the application, at time of publication,
a single Android App Bundle (AAB) is generated from the multiple CPU-specific
APKs. This AAB contains binaries for all platforms that can be uploaded to an

> app store.

Reviewer 4: It's actually the reverse: the app store generates single-architecture APKs from a multi-architecture AAB, It does this purely to save bandwidth and device storage space: all devices are capable of accepting multi-architecture APKs, and will simply ignore all the architectures other than their own.

For simple apps a multi-architecture APK and an AAB are basically equivalent. But the AAB format has additional support for splitting the app into multiple parts and distributing them independently, e.g. one part per level of a game. But none of this is really relevant to Python, so beyond mentioning the fact that both formats exist, I don't think this page needs to go into any detail about the distinction between them.

# Platforms with multiple CPU architectures

> In addition to any ABI requirements, a binary is compiled for a CPU
> architecture. That CPU architecture defines the CPU instructions that can be
> issued by the binary.

Reviewer: I think this is less clear than it could be. From my POV, anything baked into a compiled artefact is part of the ABI, including assumptions of what CPU architecture instruction set extensions are made use of. I'd say:

```suggestion
One important subset of ABI concerns is which CPU architecture a produced
binary artefact is expecting, resp. which CPU instructions from that
architecture the artifact has been compiled to use[^1]. Trying to use a binary
on a CPU not matching the target architecture specified during compilation
(or similarly, on a processor that does not support all of the instructions
the binary artifact expects to be available) will generally lead to crashes.

[^1] compare how the x86-64 architecture has various well-known extensions like SSE, SSE2, SSE3, AVX, AVX2, AVX512, etc.
```

## Current state

Historically, it could be assumed that an executable or library would be
compiled for a single CPU archicture. On the rare occasion that an operating
system was available for mulitple CPU architectures, it became the
responsibility of the user to find (or compile) a binary that was compiled for
their host CPU architecture.

However, we now see operating system platforms where multiple CPU architectures
are supported:

* In the early days of Windows NT, both x86 and DEC Alpha CPUs were supported
* Windows 10 supports x86, x86-64, ARMv7 and ARM64; Windows 11 supports x86-64
  and ARM64.
* Although Linux started as an x86 project, the Linux kernel is now available a
  wide range of other CPU architectures, including ARM64, RISC-V, PowerPC, s390

>   and more.

Reviewer: Can you think of an architecture (by any other vendor) which is _not_ supported by linux?

Author: Not off the top of my head. I'm sure one exists, but it's likely very niche and/or proprietary.

Reviewer: My point was that this sells Linux quite a bit short - the "x86 project" started 30 years ago, and for the most part of its history, it has supported every architecture under the sun that has a larger-than-microscopic amount of users.

```suggestion
* Due to its open source nature, Linux tends to support all CPU architectures for
  which someone is interested enough to author & provide support in the kernel,
  see [here](https://en.wikipedia.org/wiki/List_of_Linux-supported_computer_architectures).
```

Author: I mean... sure. Not sure it advances the discussion much, but 🤷

Reviewer 2: It's not supposed to advance the discussion, merely avoid arbitrarily volunteering falsehoods (which might result in taking the discussion *backwards*).

Reviewer: > Not sure it advances the discussion much, but 🤷

From my side, much of the discussion has wrapped up, this was a minor remaining editorial point.

* Apple transitioned Mac hardware from PowerPC to Intel (x86-64) CPUs, providing
  a forwards compatibility path for binaries
* Apple is currently transitioning Mac hardware from Intel (x86-64) to
  Apple Silicon (ARM64) CPUs, again providing a forwards compatibility
  path
* Apple supports ARMv6, ARMv7, ARMv7s, ARM64 and ARM64e on iOS
* Android currently supports ARMv7, ARM64, x86, and x86-64; it has historically
  also supported ARMv5 and MIPS

CPU architecture compatibility is a necessary, but not sufficient criterion for
determining binary compatibility. Even if two binaries are compiled for the same
CPU architecture, that doesn't guarantee [ABI compatibility](abi.md).

> In some respects, CPU architecture compatibility could be considered a superset
> of [GPU compatibility](gpus.md). When dealing with multiple CPU architectures,
> there may be some overal with the solutions that can be used to support GPUs in
> native binaries.

Reviewer: typo
```suggestion
In some respects, CPU architecture compatibility could be considered a superset
of [GPU compatibility](gpus.md). When dealing with multiple CPU architectures,
there may be some overlap with the solutions that can be used to support GPUs in
native binaries.
```

I also don't see how you mean "superset". GPU architectures are fewer and more homogeneous than CPU architectures. Or do you mean how CUDA is encouraging fat binaries for compatibility (and avoiding the JIT-overhead if it's not pre-compiled)?

Author: I mean a "superset" in the sense of that it's all the same problems/restrictions that exist for CPU compatibility, plus some more.

Three approaches have emerged on operating systmes that have a need to manage
multiple CPU architectures:

### Multiple binaries

The minimal solution is to distribute multiple binaries. This is the approach
that is by Windows and Linux. At time of distribution, an installer or other
downloadable artefact is provided for each supported platform, and it is up to
the user to select and download the correct artefact.

### Archiving

The approach taken by Android is very similar to the multiple binary approach,
with some affordances and tooling to simplify distribution.

> By default Android projects use Java/Kotlin, which produces platform independent
> code. However, it is possible to use non Java/Kotlin libraries by using JNI and
> the Android NDK (Native Development Kit). If a project contains native code, a
> separate compilation pass is performed for each architecture.

Reviewer: Please add JNI / NDK to the glossary (and other acronyms from these pages that you find useful).
```suggestion
By default Android projects use Java/Kotlin, which produces platform independent
code. However, it is possible to use non-Java/Kotlin libraries by using JNI and
the Android NDK (Native Development Kit). If a project contains native code, a
separate compilation pass is performed for each architecture.
```

   allow the user to specify a single compilation command, and instruct the
   compiler to generate multiple output architectures in the output binary
2. **Multiple pass** After compiling a binary for each platform, Apple provides
   a call named `lipo` to combine multiple single-architecture binaries into a
   single fat binary that contains all platforms.

At runtime, the operating system loads the binary slice for the current CPU
architecture, and the linker loads the appropriate slice from the fat binary of
any dynamic libraries.

On macOS ARM hardware, Apple also provides Rosetta as a support mechanism; if a
user tries to run an binary that doesn't contain an ARM64 slice, but *does*
contain an x86-64 slice, the x86-64 slice will be converted at runtime into an
ARM64 binary. Complications can occur when only *some* of the binary is being
converted (e.g., if the binary being executed is fat, but a dynamic library
isn't).

To support the transition to Apple Silicon/M1 (ARM64), Python has introduced a
`universal2` architecture target to support . This is effectively a "fat wheel"
format; the `.dylib` files contained in the wheel are fat binaries containing

> both x86_64 and ARM64 slices.

Reviewer: Missing word after "to support"

iOS has an additional complication of requiring support for mutiple *ABIs* in
addition to multiple CPU archiectures. The ABI for the iOS simulator and
physical iOS devices are different; however, ARM64 is a supported CPU
architecture for both. As a result, it is not possible to produce a single fat
library that supports both the iOS simulator and iOS devices. Apple provides an
additional structure - the `XCFramework` - as a wrapper format for packaging
libraries that need to span multiple ABIs. When developing an application for
iOS, a developer will need to install binaries for both the simulator and
physical devices.

## Problems

At present, the Python ecosystem almost exclusively uses the "multiple binary"
solution. This serves the needs of Windows and Linux well, as it matches the
way end-users interact with binaries.

> The `universal2` "fat wheel" solution also works well for macOS. The definition
> of `universal2` is a hard-coded accomodation for one specific (albiet common)

Reviewer: ```suggestion
The `universal2` "fat wheel" solution also works well for macOS. The definition
of `universal2` is a hard-coded accomodation for one specific (albeit common)
```

multi-architecture configuration, and involves a number of specific
accomodations in the Python ecosystem (e.g., a macOS-specific architecture
lookup scheme).

Supporting iOS requires supporting between 2 and 5 architectures (x86_64 and
ARM64 at the minimum), and at least 2 ABIs - the iOS simulator and iOS device

> have different (and incompatible) binary ABIs. At runtime, iOS expects to find a
> single "fat" binary for any given ABI. iOS effectively requires an analog of

Reviewer: "fat binary for a given ABI" is confusing IMO
```suggestion
have different (and incompatible) binary ABIs. At runtime, iOS expects to find a
single "fat" binary containing the necessary artifacts per required ABI.
iOS effectively requires an analog of [...]
```

(depending on the range of development and end-user configurations the app needs
to support). Android's archiving-based approach can be mapped onto the "multiple
binary" approach, as it is possible to build a single archive from multiple
individual binaries. However, some coordination is required when installing
multiple binaries. If an independent install pass (e.g., call to `pip`) is used
for each architecture, the dependency resolution process for each platform will
also be independent; if there are any discrepancies in the specific versions
available for each architecture (or any ordering instabilities in the dependency
resolution algorithm), it is possible to end up with different versions on each
platform. Some coordination between per-architecture passes is therefore
required.

## History

[The BeeWare Project](https://beeware.org) provides support for building both
iOS and Android binaries. On both platforms, BeeWare provides a custom package
index that contains pre-compiled binaries
([Android](https://chaquo.com/pypi-7.0/);
[iOS](https://anaconda.org/beeware/repo)). These binaries are produced using a

> forge-like set of tooling

Reviewer: "forge-like" unclear here as well

# Platforms with multiple CPU architectures

One important subset of ABI concerns is the CPU architecture for which a binary
artefact has been built. Attempting to run a binary on hardware that doesn't
match the CPU architecture (or architecture variant[^1]) for which the binary
was built will generally lead to crashes, even if the ABI being used is
otherwise compatible.

> 

Reviewer 3: I think we're back to the key problem being support for iOS and Android (the one other concrete case of aarch64 variants like SBSA may not be an issue after all); so tweaking the intro here would be good I think to state that clearly. We don't have concrete problems for any other platforms, because single-arch wheels are always preferred.

Author: I disagree. macOS is also firmly in this grouping, at least for my use case. 

This isn't something that affects the "Conda/System python" use case, because you're in control of the Python interpreter, and as the person installing a binary package, you're pulling binaries that are specific to your usage requirement, which is a single CPU architecture by definition.

However, if you're building a redistributable macOS app, you *must* care about all the architectures that your user base may be using. If I build an app on M1, and give it to a user on x86, I still want my app to run. The set of problems here are identical to those on mobile platforms. The Python ecosystem has *a* solution - universal wheels - but this solution is (a) not inherently useful for mobile platforms without significant changes, and (b) not liked by some segments of the desktop user base.

From a packaging perspective, I distinctly *do not* prefer single arch wheels, as they complicate the process of building an app - at least, it does given the current state of Python packaging tooling. If the Python ecosystem is going to converge on single architecture wheels as the only/preferred approach, then I'd argue that needs to be combined with a clear understanding of how use cases on platforms that need to support multiple architectures interact with that ecosystem. Being able to do "pip install" specifying multiple architectures and implicitly do a "delocate-fuse" (or equivalent) for each target architecture seems like a bare minimum.

Author: I guess I'm unclear why this is a `py2app` (or, in my case, `briefcase`) issue (beyond the obvious "it's your problem, so fix it" answer). As discussed further up the page, this isn't something that can be trivially resolved by end-user tooling. It's something that requires policies and annotation at the level of the package ecosystem - at the very least to describe how package merging *should* occur, if not to provide at least a reference implementation of that merging process.

Reviewer 3: Agreed on the improvements needed discussed in the other thread.

> I guess I'm unclear why this is a `py2app` (or, in my case, `briefcase`) issue (beyond the obvious "it's your problem, so fix it" answer).

Beyond the "those who need it tend to have to do the work", it is because from a holistic design perspective it is the correct place to implement this. The only users that need this are `py2app`, `briefcase` & co, so let's support the use case there rather than both every single package author with more work. It also will scale better in practice that way (only 50% of package authors uploading fat wheels won't be enough for your needs here).

Reviewer 3: >  It should be shared first-class tooling in the Python ecosystem that can be utilised by anyone that needs it.

I agree, and I like this. Let's state it like this in the "Potential solutions" section?

Ideally what should happen here is that this ends up in `auditwheel` and that becomes the one-stop shop for wheel distribution/packager needs, rather than having separate per-OS projects with patchy support for this kind of thing.

Reviewer 3: I've added an entry under Potential solutions for this. Please resolve if that looks good.

[^1]:
    E.g., the x86-64 architecture has a range of well-known extensions, such as
    SSE, SSE2, SSE3, AVX, AVX2, AVX512, etc.

## Current state

Historically, it could be assumed that an executable or library would be
compiled for a single CPU architecture. On the rare occasion that an operating

> system was available for mulitple CPU architectures, it became the

Reviewer 3: This sentence still needs a rewrite, an OS being supported on multiple CPU architectures is the rule rather than the exception.

Author: I was referring more to the macOS situation of user-space actively expecting multiple architectures, rather than the more general "Linux is available on almost everything"; however, I agree this is ambiguous. I'll update the text.

are supported:

* In the early days of Windows NT, both x86 and DEC Alpha CPUs were supported
* Windows 10 supports x86, x86-64, ARMv7 and ARM64; Windows 11 supports x86-64
  and ARM64.
* Due to its open source nature, Linux tends to support all CPU architectures for
  which someone is interested enough to author & provide support in the kernel,
  see [here](https://en.wikipedia.org/wiki/List_of_Linux-supported_computer_architectures).
* Apple transitioned Mac hardware from PowerPC to Intel (x86-64) CPUs, providing
  a forwards compatibility path for binaries
* Apple is currently transitioning Mac hardware from Intel (x86-64) to
  Apple Silicon (ARM64) CPUs, again providing a forwards compatibility
  path
* Apple supports ARMv6, ARMv7, ARMv7s, ARM64 and ARM64e on iOS
* Android currently supports ARMv7, ARM64, x86, and x86-64; it has historically
  also supported ARMv5 and MIPS

CPU architecture compatibility is a necessary, but not sufficient criterion for
determining binary compatibility. Even if two binaries are compiled for the same
CPU architecture, that doesn't guarantee [ABI compatibility](abi.md).

> In some respects, CPU architecture compatibility could be considered a superset

Reviewer 3: I'm not sure that this paragraph, and this sentence in particular, helps. I suspect users may think "really, how?" and there is no answer as of today because there are no solutions for GPU support at all yet, and it's unclear where that will go in the future.

Author: It was more of an idle thought than a particularly deep insight, so if you think it's more confusing than help, I'll delete it.

convention exists for providing a binary for each platform, with the same
library name.

The final binary artefact produced for Android distrobution uses this same
directory convention. A "binary" on Android is an APK (Android Application
Package) bundle; this is effectively a ZIP file with known metadata and
structure; internally, there are subfolders for each supported CPU architecture.
This APK is bundled into AAB (Android Application Bundle) format for upload to
an app store; at time of installation, a CPU-specific APK is generated and
provided to the end-user for installation.

### Fat binaries

Apple has taken the approach of "fat" binaries. A fat binary is a single
executable or library artefact that contains code for multiple CPU
architectures.

Fat binaries can be compiled in two ways:

1. **Single pass** Apple has modified their compiler tooling with flags that
   allow the user to specify a single compilation command, and instruct the
   compiler to generate multiple output architectures in the output binary
2. **Multiple pass** After compiling a binary for each platform, Apple provides
   a call named `lipo` to combine multiple single-architecture binaries into a

>    single fat binary that contains all platforms.

Reviewer 3: Note to self: add a bit here about `delocate-fuse`, which makes it easy to do this.

Author: TIL about delocate-fuse...

contain an x86-64 slice, the x86-64 slice will be converted at runtime into an
ARM64 binary. Complications can occur when only *some* of the binary is being
converted (e.g., if the binary being executed is fat, but a dynamic library
isn't).

To support the transition to Apple Silicon/M1 (ARM64), Python has introduced a
`universal2` architecture target. This is effectively a "fat wheel" format; the
`.dylib` files contained in the wheel are fat binaries containing both x86-64
and ARM64 slices.

iOS has an additional complication of requiring support for mutiple *ABIs* in
addition to multiple CPU architectures. The ABI for the iOS simulator and
physical iOS devices are different; however, ARM64 is a supported CPU
architecture for both. As a result, it is not possible to produce a single fat
library that supports both the iOS simulator and iOS devices. Apple provides an
additional structure - the `XCFramework` - as a wrapper format for packaging
libraries that need to span multiple ABIs. When developing an application for
iOS, a developer will need to install binaries for both the simulator and
physical devices.

> ## Problems

Reviewer 3: The first two paragraphs are still descriptions of the current state rather than problems. So let's move those higher up. And start here with stating that the key problem is supporting mobile platforms, and iOS and Android in particular, with their multi-architure requirements.

At present, the Python ecosystem almost exclusively uses the "multiple binary"
solution. This serves the needs of Windows and Linux well, as it matches the
way end-users interact with binaries.

> The `universal2` "fat wheel" solution also works well for macOS. The definition

Reviewer 3: I have to rewrite this - it is counter to what's more or less the consensus opinion in scientific/numerical projects I think, which is that `universal2` is not a good idea, and certainly inferior to simply distributing thin `x86_64` and `arm64` wheels.

Author: I'm intrigued about this consensus. I've definitely noticed the trend to *not* package fat binaries for numpy et al, but I haven't been clear on *why* it's been necessary. If you're able to build an x86_64 wheel and an M1 wheel, I'm not aware of any technical reason that you couldn't produce a universal2 wheel - even if it's done as a "make 2 wheels and merge".

Reviewer 2: Why does it have to (or even should be expected to) be a technical reason? There are many things that are technologically possible, that doesn't always make them a good idea.

Specifically, though, the consensus is that they are a bad user experience. For projects which are often already very large, it's seen as burdensome to make users download lots of data they don't want or need. If you have two thin wheels, it's also extra work you didn't ask for, to fuse them into universal2. Extra work which seemingly serves no purpose other than to waste the network bandwidth and disk storage space of users.

This isn't great even for small projects. SciPy is "only" 30mb. For projects that are hundreds of mb, the user experience would become prohibitively bad.

After paying all this inconvenience and wasted power/resources, what benefit is seen at the end? Do the wheels run better on those systems?

Author: Do they run better? For me, that's an unequivocal "yes, they do", because I can download a single macOS wheel with a single `pip install` command, and my macOS app will run on all Macs.

If the python/pip ecosystem had better support for multi-architecture deployment, then sure - single architecture wheels could be a better solution, requiring less download volume. However, absent that tooling, universal2 wheels provide a *much* better user experience for my users.

Reviewer 3: > I've definitely noticed the trend to not package fat binaries for numpy et al, but I haven't been clear on why it's been necessary.

There's multiple reasons:
1. The tradeoff between download size and disk space usage vs. the upside for say a .dmg installer is (a) bad (we're talking 100s of MBs per Python env, and users/devs are likely to have quite a few envs, so this is GBs wasted - and macbook disk space isn't cheap ...), and (b) really does not exist for wheels or libraries to begin with - it's only applicable to end users downloading a .dmg and then not having to choose (which means, it applies mostly to non-technical end users only)
2. It imposes an extra burden on CI and PyPI total size usage for each project
3. When you anyway provide thin wheels, you cannot even install a universal2 wheel with `pip` from PyPI at all. Why upload artifacts you cannot install?

The reasoning given for universal2 wheels is extremely thin. There's no argument there that wouldn't equally apply to, say, 32-bit and 64-bit Windows wheels. Users get confused there too, more than on macOS even. But we don't stuff 32/64-bit builds into a single Windows wheel either. Nor for any other platform. The whole `universal2` thing is just bad and should probably never have been permitted on PyPI. The _only_ folks who _maybe_ need it are people building installers directly for nontechnical end users. And those can use `delocate-fuse` easily when building those installers. 

> If the python/pip ecosystem had better support for multi-architecture deployment,

Clearly you're having a problem here, but it seems like that's specific to iOS. For macOS laptop/desktop there should never be a need for multi-architecture deployment based on wheels.

Reviewer 2: > Do they run better? For me, that's an unequivocal "yes, they do", because I can download a single macOS wheel with a single pip install command, and my macOS app will run on all Macs.

But they don't run better, even for you -- they run exactly the same.

What's better isn't running the wheels, only using developer tooling to craft those wheels into apps. End users don't transfer wheels between Macs, they only transfer apps.

> If the python/pip ecosystem had better support for multi-architecture deployment, then sure - single architecture wheels could be a better solution, requiring less download volume. However, absent that tooling, universal2 wheels provide a much better user experience for my users.

Given a choice between CPython developing tooling for multi-architecture deployment, or multi-architecture compilation, it baffles me that the powers that be chose the latter.

Reviewer 2: This is "probably okay" if the goal is just to bake the binaries into an app, but not provide that app with the ability to compile user code. Of course the python ecosystem doesn't have a clear distinction between runtime and development code...

Reviewer 3: @isuruf isn't that just a bug that was present because of `universal2` existing at all, and it's now fixed and should work just fine for fused thin wheels? See https://github.com/numpy/numpy/blob/main/numpy/core/include/numpy/numpyconfig.h#L6-L55 - the hardcoding is always there.

Second comment I have is that if one would only fuse the final artifacts that go into an app, something like this would never be a problem.

Reviewer 5: Yes, it was fixed in `numpyconfig.h`. (`_numpyconfig.h` with the `_` prefix are different for the two architectures and the lines you mentioned fixes the issue with the divergence in `_numpyconfig.h`)

> Second comment I have is that if one would only fuse the final artifacts that go into an app, something like this would never be a problem.

In this particular case, yes it is not an issue because numpy's runtime behavior does not depend on the header file. However a package at runtime might depend on a generated header or a generated configuration file and it might be an issue. I haven't seen any package like that at the moment. Just an issue to keep in mind.

Author: > > It's probably a cultural thing
> Part of the issue is that Apple itself has been pushing for this kind of thing. Perhaps there's even a business perspective there that's anti-aligned with end user interest (Apple probably doesn't care about disk space usage, maybe they even sell beefier machines).

I strongly suspect that this is is indeed a cultural issue - but not the one you think it is.

Most users of computers *do not not know or care* what CPU their computer has. It is a utility device for getting a job done. To the extent that they are aware of CPU architecture during a laptop purchase, it will be a bullet point on the sales chart next to "rose gold unibody aluminium frame". 

In this environment, Apple has made the decision to trade off disk space (which is relatively cheap, plentiful, user-extensible, and recoverable) for user convenience. The user doesn't need to know or care what CPU they have - well behaved Apple apps run on *every* Apple machine.

Compare this with the Windows on ARM experience, where just this weekend, one of the attendees in the tutorial I was running was having difficulties, and it wasn't until I asked "wait - is that an ARM surface" that it occurred to him that this could be the source of the problems he was having. This is a class of user confusion that Mac users *do not have*. This was at a conference for Open Source enthusiasts, so it's hard to push the argument that this was a "non technical  user" problem.

Universal2 isn't the *only* solution to this problem. Having single architecture wheels *and* core packaging infrastructure that allows for installing multiple platform wheels for the same install would also be a viable solution to this problem. However, at present, the former exists, and the latter doesn't.

Author: > One thing to keep in mind is that merging two thin wheels might not make a universal2 with the desired functionality. For eg: numpy generated header files with the size of `long double` hardcoded and running `delocate-fuse` would pick the header from `x86_64` wheel which was wrong for the `arm64` build. (This was fixed by generating a header that works for both architectures.)

On the same theme - both wheels will contain a full copy of the Python code for the package. When you merge 2 binary wheels, which copy of the code do you keep? If there's any sort of conditional module inclusion in the packaging process (e.g., only including an `arm64.py` or `x86_64.py` depending on the target platform), or any sort of build-time (re)writing of modules (e.g., renaming `arm64.py` to `engine.py` as part of the build, those changes *also* need to be merged. 

I'm sure it's possible to come up with policy rules that allow for this sort of thing, but it's something that needs to be explicit.

Reviewer 2: > When you merge 2 binary wheels, which copy of the code do you keep?

I would say: validate that the files are the same, and if they're not the same, error out. This problem exists whether universal2 or thin wheels are being built, though. Maybe it's a bit *easier* to catch with fusing thin wheels, however, since you'd sidestep the chance the build system will produce fat binaries but do pure-python generation based on whatever the current running platform seems to be.

Reviewer 3: This merging strategy is a very good point indeed. It would have caught the bug that @isuruf pointed out. I think there's a bit more to it than "error out"; that'd be a good default but doesn't always work. E.g., numpy/scipy have a generated `__config__.py` containing build-time paths. The merge strategy there should be "random" not "error". If that could be specified per file, it'd also be good for the reproducible builds effort.

> Most users of computers _do not not know or care_ what CPU their computer has.

I agree. That doesn't translate to "users of wheels" though. 

> In this environment, Apple has made the decision to trade off disk space (which is relatively cheap, plentiful, user-extensible, and recoverable) for user convenience.

I'd say that that's a value judgement. It hasn't applied for me in the past - when I was a student and recent graduate without much money, I bought the minimum spec'd machine (128 GB disk, 4/8 GB RAM IIRC) and usually had to manage disk space very carefully. I still hear these complaints from users that buy a MacBook Air. A few hundred dollars more for a larger disk is a lot of money for many demographics. Same for bandwidth - we no longer care too much in our privileged positions, but in most countries in this world, unlimited plans are not common or not even available. So my value judgement here is: `universal2` wheels are quite dismissive of those concerns, and against the inclusiveness the Python community aims for.

> Having single architecture wheels and core packaging infrastructure that allows for installing multiple platform wheels for the same install would also be a viable solution to this problem. However, at present, the former exists, and the latter doesn't.

Both exist and both are incomplete I'd say (because you can't even install `universal2` wheels with `pip` when all thin/fat flavors are uploaded to PyPI - it seems really no one uses or cares much about `universal2` beyond a handful of people). Right now the `delocate-fuse` approach is _more_ incomplete, so let's add as a solution direction to address that?

Reviewer 2: Oh. Speaking of which -- Apple made the curious decision to make storage space not, in fact, extensible. ;)

The hard drive is soldered in, you can't replace it with a larger capacity after buying the device. So much for user convenience.

If we're discussing why Apple makes decisions, I would suspect the real reason Apple made the universal2 decision is to get people to buy the comparatively expensive luxury devices with high-specced storage while simultaneously dodging the blame by making it seem like something you just do because of software distributors.

Reviewer 3: @freakboy3742 I pushed a commit which adds a foldout with title _"what's the deal with `universal2`?"_. I provided arguments against; if you could add arguments in favor, that'd be great. Once we have that, we can update the current state and problem sections.

of `universal2` is a hard-coded accomodation for one specific (albeit common)
multi-architecture configuration, and involves a number of specific
accomodations in the Python ecosystem (e.g., a macOS-specific architecture
lookup scheme).

Supporting iOS requires supporting between 2 and 5 architectures (x86-64 and
ARM64 at the minimum), and at least 2 ABIs - the iOS simulator and iOS device
have different (and incompatible) binary ABIs. At runtime, iOS expects to find a
single "fat" binary for the ABI that is in use. iOS effectively requires an
analog of `universal2` covering the 2 ABIs and multiple architectures. However:

1. The Python ecosystem does not provide an extension mechanism that would allow

>    platforms to define and utilize multi-architecture build artefacts.

Reviewer 3: Side note: there is already a way to sneak multiple CPU architecture tags into a single wheel filename, because that's what manylinux filenames contain:
```
numpy-1.24.2-pp38-pypy38_pp73-manylinux_2_17_x86_64.manylinux2014_x86_64.whl 
```
It doesn't do anything useful though; I believe the part after the dot is simply ignored because of implementation details (and it's an alias here). So you are probably able to put in a different CPU architecture today, and it won't affect install/resolution logic. I have not found anything in any PEP that supports doing what packaging tools are doing here for manylinux.

Author: Yeah - I'm not aware of any formal standardization of this either. I'd feel a lot more comfortable if it was formalised, rather than being an exploit of an accidental edge case.

      or similar formats, produced by tools such as
      [py2app](https://py2app.readthedocs.io) or
      [briefcase](https://beeware.org/project/projects/tools/briefcase/)),
      the person building the project must be accommodate all possible CPU
      architectures where the code *could* be executed.
    - If binary wheels are only available in "thin" format, any issues with
      merging those wheels into fat equivalents for distribution purposes are
      deferred to the end user. This can be problematic as it may require
      expert knowledge about the package being merged (such as optional
      modules or header files that may not be present in both thin artefacts).
      Universal2 artefacts captures this knowledge by requiring the project
      maintainers to resolve any merging issues.

    Arguments against:

    - `universal2` wheels are never necessary for end users installing into
      a locally installed Python environment exclusively for their own use,
      which is the default experience most users have with Python.
    - The tradeoff between download size and disk space usage vs. the upside

>       for say a .dmg installer is bad - for a typical PyData stack it takes

Author: "The tradeoff... is bad" is highly subjective language.

Reviewer 3: Isn't this already implied in the "for" and "against" labels above the lists? I'd expect your "for" argument to say kinda the opposite. It's a tradeoff, so how you make it is inherently subjective. Happy to rephrase though, but not sure how. What would you suggest?

Author: "Using universal2 wheels requires larger downloads and more disk space" is undeniably true. "Using universal2 requires a tradeoff between download size/disk space and user experience" is true for at least a subset of users. "This tradeoff is bad" is a value judgement. As I've made clear a couple of times in this thread, I consider the tradeoff of disk space to be entirely acceptable, and not "bad" at all.

Reviewer 2: :thinking:

Arguments for:
- the tradeoff is good because XXX

Arguments against:
- the tradeoff is bad because YYY

The current wording in this bullet point is not saying "the tradeoff is bad", it's saying "here is a persuasive argument for why I assert that it's bad". As such, I argue that it's not subjective language, it's a short persuasive essay.

Reviewer 3: Okay, saying "using universal2 wheels requires larger downloads and more disk space" followed by the "- for a typical PyData stack" is fine with me here.

      Meaning that defaulting to `universal2` would use several GBs of disk
      space more.

        - Disk space on older MacBook Air models is 128 GB, and up to half of that
          can be taken up by the OS and system data itself. So a few GBs can be
          significant.
        - Internet plans in many countries are not unlimited; almost doubling the
          download size of wheels is a serious cost, and not desirable for any
          user - but especially unfriendly to users in countries where network
          infrastructure is less developed.

    - In addition, it takes extra space on PyPI (examples: `universal2` wheels
      cost an extra 81.5 MB for NumPy 1.21.4 and 175.5 MB for SciPy 1.9.1), and
      projects with large wheels often run into total size limits on PyPI.
    - It imposes an extra maintenance burden for each project, because separate
      CI jobs are needed to build and test `universal2` wheels. Typically
      projects make tradeoffs there, because they cannot support every
      platform. And `universal2` doesn't meet the bar for usage frequency /
      user demand here - it's well below the demand for `musllinux`,

>       `ppc64le`, PyPy, and other such platforms with still patchy support

Reviewer 2: I don't read that as the intent here.

Rather, the demand for universal2 is limited to the subset of M1 mac users who aren't satisfied with `pip install` of a thin wheel into a locally active environment.

The demand for ppc64le and musllinux wheels, however, is the *entire* group of users who are using ppc64le or the musl libc.

>       `ppc64le`, PyPy, and other such platforms with still patchy support

Reviewer 3: Indeed, I meant "all ppc64le or musllinux users" vs. "builders of universal apps for macOS". There really is very little demand for `universal2` as far as I can tell, while there's clearly a ton of users on macOS M1.

>       `ppc64le`, PyPy, and other such platforms with still patchy support

Author: Ok - that's a very different (and much more reasonable claim); but definitely not what I read from the way that statement is phrased.

I'll also point out there's a certain amount of post hoc ergo propter hoc here - if universal2 wheels don't exist, there's not going to be very much evidence of demand for installing them in *any* circumstance, let alone in the specific case of app distribution :-)

>       `ppc64le`, PyPy, and other such platforms with still patchy support

Reviewer 3: I'll rephase this then to make clearer that this is the comparison, vs. app builders and not M1 users.

> I'll also point out there's a certain amount of post hoc ergo propter hoc here 

I don't know about that. It's the same as musllinux & co - when we don't provide them (which we haven't so far), people come and ask. I'm pretty certain that the folks who show up as Alpine users and ask for Alpine (so musllinux) wheels has substantially outnumbered the  folks asking for universal2 wheels because they needed them for something.

>       `ppc64le`, PyPy, and other such platforms with still patchy support

Reviewer 2: This bullet point on the "arguments against" really congeals down to a matter of necessities:
- a previous bullet point asserted that universal2 is not *necessary* for a specific range of cases
- for ppc64le or musllinux, it is necessary to have single-architecture wheels for the current single architecture, there is no philosophical way around this

      (see [Expectations that projects provide ever more wheels](../../meta-topics/user_expectations_wheels.md)
      for more on that).
    - When a project provides thin wheels (which is a must-do for projects with
      native code, because those are the better experience due to smaller

>       size), you cannot even install a `universal2` wheel with pip from PyPI at

Author: Again, subjective, use-case specific language. In my use case, universal2 wheels unequivocally provide a better experience.

Reviewer 2: "because those are the better experience with pip install into an active environment" perhaps?

Because it *is* unequivocally better for all use cases where you aren't going to transfer the results to another computer. And virtual environments are not relocatable. And I'd imagine that conda doesn't provide a universal2 interpreter so conda environments *might* be relocatable, but presumably not across machines.

Reviewer: >  And I'd imagine that conda doesn't provide a universal2 interpreter so conda environments might be relocatable, but presumably not across machines.

Conda environments should be more or less relocatable (not that I'd recommend it), but only within the same CPU architecture. Everything in conda is per-arch, there are no fat binaries or something like universal2.

Reviewer 3: > "because those are the better experience with pip install into an active environment" perhaps?

That sounds fine to me. Maybe adding "as an end user"? I imagine `briefcase` and `py2app` also use venv's under the hood.

Reviewer 2: I was thinking that those would be inactive environments, but it's a subtle distinction and "as an end user" is probably a better one.

Author: FWIW, Briefcase doesn't use literal venvs, but it does do some venv-like tricks to ensure interpreter isolation. I can't speak to the current state of py2app.

Reviewer 3: I toned this down from "must do" to "should be done" and added the clarification that it's only better when installing for use on that machine. Please resolve this comment if that looks good.

>       all. Why upload artifacts you cannot install?

Author: Undeniably true with current tooling - but I'd argue this is a bug with pip, born from a lack of support for multiple architectures (which is the point this document is trying to raise), rather than the end point of the discussion.

Reviewer 3: Sure, but again - this is my "against" argument as things stand today. If this is fixed, we can also fix/standardize `delocate-fuse`. 

I could add in brackets "(this should be fixed in a future pip release, see `pip#xxx`)", does that sound fine?

Author: Note added.

    and isn't well-supported by either packaging tools (e.g., there is no way
    to install a `universal2` wheel from PyPI if thin wheels are also present)
    or package authors (most numerical, scientific and ML/AI package authors do
    not provide them). There are some arguments for and against supporting the
    format or even defaulting to it.

    Arguments for:

    - While users with a technical background are usually aware of the CPU
      architecture of their machine, less technical users are often unaware of
      this detail (or of the significance of this detail). The universal2 wheel
      format allows users to largely ignore this detail, as all CPU
      architectures for the macOS platform are accomodated in a single binary
      artefact.
    - macOS has developed an ecosystem where end users expect that any macOS
      binary will run on any macOS machine, regardless of CPU architecture.
      As a result, when building macOS apps (`.dmg` downloadable installers
      or similar formats, produced by tools such as
      [py2app](https://py2app.readthedocs.io) or
      [briefcase](https://beeware.org/project/projects/tools/briefcase/)),

>       the person building the project must be accommodate all possible CPU

Reviewer 3: minor textual: "be accommodate" -> "accommodate"

      projects with large wheels often run into total size limits on PyPI.
    - It imposes an extra maintenance burden for each project, because separate
      CI jobs are needed to build and test `universal2` wheels. Typically
      projects make tradeoffs there, because they cannot support every
      platform. And `universal2` doesn't meet the bar for usage frequency /
      user demand here - it's well below the demand for `musllinux`,
      `ppc64le`, PyPy, and other such platforms with still patchy support
      (see [Expectations that projects provide ever more wheels](../../meta-topics/user_expectations_wheels.md)
      for more on that).
    - When a project provides thin wheels (which is a must-do for projects with
      native code, because those are the better experience due to smaller
      size), you cannot even install a `universal2` wheel with pip from PyPI at
      all. Why upload artifacts you cannot install?
    - It is straightforward to fuse two thin wheels with `delocate-fuse` (a
      tool that comes with [delocate](https://pypi.org/project/delocate/)),
      it's a one-liner: `delocate-fuse $x86-64_wheel $arm64_wheel -w .`

>       However, it's worth noting that this requires that any headers or python

Reviewer 3: I don't think this edit is right. The exact same bug will still be present in universal2 wheels. Anything that depends on the build machine's architecture is going to be incorrect and must be explicitly handled for universal2 I believe. There may be exceptions, but I think it's true in general (similar issues as when cross-compiling) and it was true for the `long double` size issue that Isuru linked.

The problem here is the existence of universal2 as a format more than how you produced it.

Reviewer 3: Actually, it's a more general problem even for any multi-architecture support. Typical header generation that does things like `sizeof(<C type name>)` is going to be incorrect.

Author: Also - doesn't this *support* my point? Building a fat wheel *isn't* a "straightforward" matter of running `delocate-fuse`".

Reviewer 3: Sure, anything is possible, but it requires care. 

> Building a fat wheel _isn't_ a "straightforward" matter of running `delocate-fuse`".

I think that in the case of no bugs like architecture-dependent headers, it _is_ straightforward. And if there are such bugs, I'd rather get a loud clash from `delocate-fuse` rather than a silent bug.

In the end, I think the most robust process would be to build thin apps based on only thin wheels, and then only fuse the final thin apps together. That way, most/all bugs like this are likely irrelevant.

Author: Unfortunately, building thin apps isn't a viable option. It causes a degraded and user experience on macOS, and fat apps are a *requirement* for iOS/Android distribution.

Author: (noting that the "fat" interpretation on Android is a little easier to accomodate because of how it handles binaries)

Reviewer 3: > Unfortunately, building thin apps isn't a viable option. It causes a degraded and user experience on macOS, and fat apps are a _requirement_ for iOS/Android distribution.

This is why I said "and then only fuse the final thin apps together". I meant "into fat aoos to distribute to users". The end result is the same, a fat app. It's just that the process is better controlled; any intermediate header issues get avoided when you use a stack of thin wheels.

Reviewer 2: Your edit notes that delocate-fuse might succeed but produce broken artefacts.

But it's not an inherent problem with delocate-fuse, it's a "simple" matter of the project not supporting or testing universal2. If a project supports or tests universal2, it should not matter whether the project personally executes the compiler in multi-pass mode, or personally executes delocate-fuse.

Either both work, or both do not work.

And if/when the delocate-fuse tooling is improved, then it *does* become a straightforward matter of "attempt to fuse the wheels, if it doesn't work then submit a bug report to the project". No more "the project doesn't test this, so there may be subtle runtime bugs".

Reviewer 3: I have rephrased this now to say: 

_Note though that robustness improvements in delocate-fuse for more complex cases (e.g., generated header files with architecture-dependent content) are needed (see [delocate#180](https://github.com/matthew-brett/delocate/issues/180)). Such cases are likely to be equally problematic for direct universal2 wheel builds (see, e.g., [numpy#22805](https://github.com/numpy/numpy/pull/22805))._

I think that addressed it, please resolve if this looks good.

# Platforms with multiple CPU architectures

One important subset of ABI concerns is the CPU architecture for which a binary
artefact has been built. Attempting to run a binary on hardware that doesn't
match the CPU architecture (or architecture variant[^1]) for which the binary
was built will generally lead to crashes, even if the ABI being used is
otherwise compatible.

[^1]:
    E.g., the x86-64 architecture has a range of well-known extensions, such as
    SSE, SSE2, SSE3, AVX, AVX2, AVX512, etc.

## Current state

Most operating systems support multiple CPU architectures:

* In the early days of Windows NT, both x86 and DEC Alpha CPUs were supported
* Windows 10 supports x86, x86-64, ARMv7 and ARM64; Windows 11 supports x86-64
  and ARM64.
* Due to its open source nature, Linux tends to support all CPU architectures for
  which someone is interested enough to author & provide support in the kernel,

>   see [here](https://en.wikipedia.org/wiki/List_of_Linux-supported_computer_architectures).

Reviewer 6: Could you use an informative link label instead of `here`?

(there are other cases of this problem we can see in this PR, but good to start somewhere 🙂)

(refs https://www.w3.org/QA/Tips/noClickHere – https://www.smashingmagazine.com/2012/06/links-should-never-say-click-here/ – https://stephanieleary.com/2015/05/why-click-here-is-a-terrible-link-and-what-to-write-instead/ etc)

multiple CPU architectures:

### Multiple binaries

The minimal solution is to distribute multiple binaries. This is the approach
that is by Windows and Linux. At time of distribution, an installer or other
downloadable artefact is provided for each supported platform, and it is up to
the user to select and download the correct artefact.

### Archiving

The approach taken by Android is very similar to the multiple binary approach,
with some affordances and tooling to simplify distribution.

By default Android projects use Java/Kotlin, which produces platform independent
code. However, it is possible to use non Java/Kotlin libraries by using JNI and
the Android NDK (Native Development Kit). If a project contains native code, a
separate compilation pass is performed for each architecture.

If a native binary library is required to compile the Android application, a
version must be provided for each supported CPU architecture. A directory layout
convention exists for providing a binary for each platform, with the same
library name.

The final binary artefact produced for Android distrobution uses this same
directory convention. A "binary" on Android is an APK (Android Application

> Package) bundle; this is effectibely a ZIP file with known metadata and

Reviewer 4: ```suggestion
Package) bundle; this is effectively a ZIP file with known metadata and
```

      platform. And `universal2` doesn't meet the bar for usage frequency /
      user demand here - it's well below the demand for `musllinux`,
      `ppc64le`, PyPy, and other such platforms with still patchy support
      (see [Expectations that projects provide ever more wheels](../../meta-topics/user_expectations_wheels.md)
      for more on that).
    - When a project provides thin wheels (which is a must-do for projects with
      native code, because those are the better experience due to smaller
      size), you cannot even install a `universal2` wheel with pip from PyPI at
      all. Why upload artifacts you cannot install?
    - It is straightforward to fuse two thin wheels with `delocate-fuse` (a
      tool that comes with [delocate](https://pypi.org/project/delocate/)),
      it's a one-liner: `delocate-fuse $x86-64_wheel $arm64_wheel -w .`
    - Open source projects rely on freely available CI systems to support
      particular hardware architectures. CI support for macOS `arm64` was a
      problem at first, but is now available through Cirrus CI. And that
      availability is expected to grow over time; GitHub Actions and other

>       providers will roll out support at some point. This allows building thin

Reviewer: Tracking issue for GHA: https://github.com/github/roadmap/issues/528

    format or even defaulting to it.

    Arguments for:

    - While users with a technical background are usually aware of the CPU
      architecture of their machine, less technical end users are often
      unaware of this detail (or of the significance of this detail). The
      universal2 wheel format allows users to largely ignore this detail, as
      all CPU architectures for the macOS platform are accomodated in a single
      binary artefact.
    - macOS has developed an ecosystem where end users expect that any macOS
      binary will run on any macOS machine, regardless of CPU architecture.
      As a result, when building macOS apps (`.dmg` downloadable installers
      or similar formats, produced by tools such as
      [py2app](https://py2app.readthedocs.io) or
      [briefcase](https://beeware.org/project/projects/tools/briefcase/)),
      the person building the project must be accommodate all possible CPU
      architectures where the code *could* be executed.
    - If binary wheels are only available in "thin" format, any issues with
      merging those wheels into fat equivalents for distribution purposes are

>       deferred to the end user. This can be problematic as it may require

Reviewer 2: End user as described in this system is people installing apps, and distinguished from "person building the project".

The issues with merging the wheels into fat equivalents is surely not deferred to the end user? It is deferred to the person building the "macOS apps" project.

Author: Agreed the usage of "end user" is ambiguous here. I'll clean that up.

      expert knowledge about the package being merged (such as optional
      modules or header files that may not be present in both thin artefacts).
      Universal2 artefacts captures this knowledge by requiring the project

>       maintainers to resolve any merging issues.

Reviewer 2: Requiring the maintainers of the wheel project, not the macOS app project.
